server/src/main/java/org/elasticsearch/tasks/Task.java

```diff
@@ -64,7 +65,8 @@ public class Task implements Traceable {
         X_OPAQUE_ID_HTTP_HEADER,
         TRACE_PARENT_HTTP_HEADER,
         TRACE_ID,
-        X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER
+        X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER,
+        PROJECT_ID
```

Comment:
Do you need these changes to make the tests on this branch work? Or do you only need them to make serverless PR work? If it's the latter, maybe you could talk with @ywangd as he's preparing serverless in a multi-project setting and has a serverless branch that uses the internal multi-project branch.

Reply:
I think we should not have this change to the open repo at this stage. I haven't read the PRs. Intuitively, we should leave anything related to projects outside the main branch until the designated merge date. We might be able to extract more generic changes from the PRs so that they can be merged separately. The rest should leverage the multi-project branch of the two repositories. It does incur developement overhead which is something we have to work with right now.

Also ping @tvernum

Reply:
>Do you need these changes to make the tests on this branch work? Or do you only need them to make serverless PR work? 

It's to make the serverless PR work. Without this change the project id is not available in the thread context during saml auth. 

This is the only project specific change that is needed to make the new multi-project saml realm work in serverless. My thinking around this is that since we're going to add this soon anyway it might make sense to approach is this way. What's the reason for keeping any project related code out of the main repo? 

If we feel that this change is not acceptable here we could either: 
- Merge this to the multi-project branch - the serverless tests will fail until the project id issues is fixed but the new realm also won't be used. 
- Come up with a way that serverless can inject a list of headers not to stash during saml auth (to hide the project specific stuff)